Preferences cleanup

wolfgang42 opened the pull request:
As seen in #22, the way preferences are handled is a bit confusing and disjointed. This change puts all of the preference getters in a separate `Prefs` class, which has several benefits:

1. Removes members of `MainActivity` that were storing some of the preferences, along with the code to bundle and unbundle them.
2. Puts all of the preferences in a single place where they can easily be seen.
3. Removes code duplication where some preferences were gotten in multiple places. If these default was to be changed the getters would have had to have been altered in multiple places; with this change only the one place in the `Prefs` class needs to be altered.

I'm submitting this as a PR separate from the fix for #22 as I haven't implemented that yet and I wanted to avoid merge conflicts later on.

wolfgang42 commented:
Right after I submitted this PR I noticed the [fdw/Goodtime@refactor](https://github.com/fdw/Goodtime/commits/refactor) branch, which doesn't separate out the `Prefs` class but also has some good ideas, and which will have a massive merge conflict if this is merged. @fdw, would you care to chime in on the discussion about cleaning up the preferences? Perhaps I should close this PR and open another issue?

fdw commented:
Thanks for noticing my refactoring, and asking before I'm sitting on a massive merge conflict :)

I quite like the idea of a separate Preferences class, so that the default values are encapsulated there. This is definitely helpful. (Although I'd rather call it Preferences instead of Prefs.)

Regarding a good merging strategy: I think I'm quite done refactoring, so how about I start a pull request, and you can then rebase your changes on that? I think this would be easier than the other way around. Would that work for you?

wolfgang42 commented:
@fdw Sure, sounds good to me! I'll star work on that in a little bit.

fdw commented:
Thanks :)

You might want to take a look at `PreferenceKeys`, this should probably be merged into your `Prefs`. And maybe the `migrateSettings()` from `MainActivity` too?

wolfgang42 commented:
OK, I just rebased all my changes on top of #29. Since this PR is based on that one, it's probably a good idea to merge #29 *first,* and then this one.

Of particular note: I took @fdw's advice and renamed `PreferenceKeys` into `Preferences` before merging all of the getters in, and also moved `migratePreferencesIfNeeded()` into the `Preferences` class.

adrcotfas commented:
Hi and thanks for the great contribution! :+1: I'll merge as soon as I get home because I'm traveling at the moment.

adrcotfas commented on app/src/main/java/com/apps/adrcotfas/goodtime/MainActivity.java:
```diff
@@ -111,35 +100,29 @@ protected void onResume() {
 
-        migratePreferencesIfNeeded();
-    }
+        this.mPref.migratePreferencesIfNeeded();
```
This line is causing a crash because this.mPref is not constructed yet. It's null at this point.

adrcotfas replied:
I just removed the function completely since it's not needed anymore. I used it to avoid a crash in June but no users are using a version this old.

wolfgang42 replied:
Sorry, that was my fault. I managed to botch the rebase and then test the wrong version.

adrcotfas commented:
Hey, I updated the contributors page.
Please give thumbs up or down.
Maybe for some reason you want it spelled differently or don't want it mentioned at all :)
Commit: https://github.com/adrcotfas/Goodtime/commit/d5036f1da8443e205246a3d9ec27c4e3339175a9
